### PR TITLE
chore: Replace `flatten`/`unflatten` by native nesting

### LINKED ISSUE #1868: New implementation of accessors to the fields using the dot notation
Ours configs require accessing nested fields using "dot" notation, e.g. the fields encoded as `a.b.c[0][1][3]`.

Currently it is handled in [`flatten.rs`](https://github.com/timberio/vector/blob/master/src/event/flatten.rs)/[`unflatten.rs`](https://github.com/timberio/vector/blob/master/src/event/unflatten.rs), but a different implementation is needed for fully nested structures.

I plan to implement approach similar to the one used in Lodash's [`_.get`](https://github.com/lodash/lodash/blob/master/get.js)/[`_.set`](https://github.com/lodash/lodash/blob/master/set.js), namely:

1. Create type `PathNode` defined as
    ```rust
    enum PathNode {
        Key(String),
        Index(usize),
    }
    ```
2. Create a function which would take a string like `a.b.c[0][1][3]` and produce an iterator over `PathNode` values. The most reasonable way to do it is to use a state machine.
3. Implement `insert`, `get`, `remove`, `all_fields` functions which would take an iterator over `PathNode` values and perform the corresponding action on a mutable reference to `BTreeMap<Atom, Value>`.
4. Use the functions listed above to reimplement `insert`, `get`, `remove`, `all_fields` methods of `Event`.

All this should preserve the current functionality, but utilize the new nested field structure.

### PR DESCRIPTION
**Pull request by a-rodin**
Closes #1868, ref #704, ref  https://github.com/timberio/vector/pull/1662#issuecomment-587362041, ref https://github.com/timberio/vector/pull/1825.

This doesn't change any documented behavior observable by the user, so no documentation changes are necessary.

Further improvements can include replacement of flat `insert()/get()/...` to their direct counterparts which would not attempt to walk the event recursively in the parts of code where such a replacement can be done, but I think it should be done as a part of one of the "tech debt" milestones.

### CODE REVIEW

File: src/event/util/log/all_fields.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+enum Node<'a> {
+    Key(&'a Atom),
+    Index(usize),
+}
+
+#[derive(Clone)]
+struct FieldsIter<'a> {
+    stack: Vec<LeafIter<'a>>,
+    nodes: Vec<Node<'a>>,
+}
```
Comment (lukesteensen) [inline src/event/util/log/all_fields.rs]: Could you add a high-level comment explaining roughly the algorithm here? The use of `stack` is relatively straightforward for a depth-first traversal, but I'm a little less clear on how we're using `nodes`.

File: src/event/util/log/all_fields.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+enum LeafIter<'a> {
+    Map(btree_map::Iter<'a, Atom, Value>),
+    Array(iter::Enumerate<slice::Iter<'a, Value>>),
+}
+
+#[derive(Clone)]
+enum Node<'a> {
+    Key(&'a Atom),
+    Index(usize),
+}
```
Comment (lukesteensen) [inline src/event/util/log/all_fields.rs]: I'm trying to think if there's a more descriptive name we could use for this. `Node` makes me think of something like a linked list node that holds a value, whereas this is not quite that. Maybe something like `PathComponent`? That feels more precise, but it is unfortunately long.

**Comment by binarylogic** [comment]
@a-rodin does it make sense for me to update documentation that discussing nested fields? Ex: The nested example here: https://vector.dev/docs/about/data-model/log/#examples.

**Comment by a-rodin** [comment]
@binarylogic This PR doesn't change any observable behavior related to nested fields, so I think there is no need to update it. But we might need to update it after the `lua` transform gets support for nesting.

File: src/event/util/log/all_fields.rs (hunk 2 above)
Comment (a-rodin) [reply src/event/util/log/all_fields.rs]: It was changed to `PathComponent`.

File: src/event/util/log/all_fields.rs (hunk 1 above)
Comment (a-rodin) [reply src/event/util/log/all_fields.rs]: The `nodes` name was changed to `path` and a comment was added.

**Pull request merged**